packages/langium/src/utils/grammar-util.ts

```diff
@@ -54,26 +54,38 @@ export function getAllReachableRules(grammar: ast.Grammar, allTerminals: boolean
     }
 
     const topMostRules = [entryRule as ast.AbstractRule].concat(getHiddenRules(grammar));
+    const rules = new Set<ast.AbstractRule>();
     for (const rule of topMostRules) {
-        ruleDfs(rule, ruleNames, allTerminals);
+        ruleDfs(rule, ruleNames, rules, allTerminals);
     }
 
-    const rules = new Set<ast.AbstractRule>();
+    
     for (const rule of grammar.rules) {
-        if (ruleNames.has(rule.name) || (ast.isTerminalRule(rule) && rule.hidden)) {
+        if ((ast.isTerminalRule(rule) && rule.hidden)) {
             rules.add(rule);
         }
     }
     return rules;
 }
 
-function ruleDfs(rule: ast.AbstractRule, visitedSet: Set<string>, allTerminals: boolean): void {
-    visitedSet.add(rule.name);
+function ruleDfs(rule: ast.AbstractRule, visitedRuleNames: Set<string>, visitedRules: Set<ast.AbstractRule> , allTerminals: boolean): void {
+    visitedRuleNames.add(rule.name);
```

Comment:
Given that we don't allow rule overrides, I don't think we need this set?

Reply:
yes